backend/downsample.py

```diff
@@ -0,0 +1,260 @@
+        max_records: An interger representing number of records to save per second.
+
+    Returns:
+        A list of downsampled records with number under max_records.
+    """
+    if strategy == 'max':
+        res = _max_min_downsample(
+            records, is_max=True, max_records=max_records)
+    elif strategy == 'min':
+        res = _max_min_downsample(
+            records, is_max=False, max_records=max_records)
+    elif strategy == 'avg':
+        res = _average_downsample(
+            records, max_records=max_records)
+    elif strategy == 'lttb':
+        res = _lttb_downsample(
+            records, max_records=max_records)
+    else:
+        res = list()
+    return res
+
+
+def downsample(filename, strategy, max_records_per_second):
+    """Reads the raw data file and downsample with the given strategy.
+
+    Assume the records file is on local disk, read the records and
+    downsample the records in one-second-at-a-time fashion.
+
+    Args:
+        filename: A string representing name of the records file.
+        strategy: A string representing downsampling strategy.
+        max_records_per_second: An interger representing number of records to save per second.
+
+    Returns:
+        A list of the downsampled data in the given file.
+    """
+    data = list()
+
+    with open(filename, 'r') as filereader:
+        temp_store = list()
+        for line in filereader:
+            temp_store.append(utils.parse_csv_line(line))
+            if temp_store[-1][0] - temp_store[0][0] >= SECOND_TO_MICROSECOND:
+                downsampled_records = _strategy_reducer(
+                    temp_store, strategy, max_records_per_second)
+                data.extend(downsampled_records)
+                temp_store = list()
+        last_second_records = _strategy_reducer(
+            temp_store, strategy, max_records_per_second)
+        data.extend(last_second_records)
+    return data
+
+
+def secondary_downsample(filename, strategy, max_records, start, end):
+    """Reads the preprocessing file and downsample with the given strategy for HTTP request.
+
+    Assume the records file is on local disk, read the records and downsample the records
+    to be within max records.
+    Optional arguments start and end to specify a timespan in which records must be laid.
+
+    Args:
+        filename: A string representing name of the records file.
+        strategy: A string representing downsampling strategy.
+        max_records: An interger representing number of records to save.
+        start: An interger representing start of timespan.
+        end: An interger representing the end of timespan.
+
+    Returns:
+        Downsampled data in the given file.
+    """
+
+    with open(filename, 'r') as filereader:
+        data = list()
+        for line in filereader.readlines():
+            record = utils.parse_csv_line(line)
+            if start is None or end is None or start <= record[0] <= end:
```

Comment:
You are right, I didn't even consider the case of only one of start or end is None, because the frontend would not send request like this...
But considering integrity of this function, it's good to have support for one-sided time limit.

When the chart is asking for data for a new strategy or the chart is starting up, we should give downsampled data in the entire time series, so no start and end are given in this case.